### PR TITLE
Working spectral partitioning in Python

### PR DESCRIPTION
**Pull request by tinesubic**
image

### CODE REVIEW

File: src/modules/partitioning/k_means_sphere.py [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+            # recompute centroids
+            for clustN in range(0, k):
+                assigned_idxs = assignment == clustN
+
+                if assigned_idxs.astype(dtype=int).sum() > 0:
+                    Yn = X[assigned_idxs, :]
+                    C[clustN, :] = calc_centroid(Yn)
+                else:
+                    C = kmeanspp(X, k)
```
Comment (lstopar) [inline src/modules/partitioning/k_means_sphere.py:104]: The procedure when losing a centroid should be updated with the current MATLAB code. Essentially, when losing a centroid you should re-run kmeanspp

File: src/modules/partitioning/spectral.py [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        cut_sum = cut_intensity(Q, part, k)
+        cut_edges = cut_size(Q, part, k)
+
+        avg_cut_intens = cut_sum / cut_edges
+
+        rnd_part = 1 + np.floor(k * np.random.rand(n))
+        rnd_cut_sum = cut_intensity(Q, rnd_part, k)
+        avg_rnd_cut_intens = rnd_cut_sum / cut_edges
+        rnd_cut_edges = cut_size(Q, rnd_part, k)
+        mean_rnd_cut_intens = mean_rnd_cut_intens + (avg_rnd_cut_intens / total_tries)
```
Comment (lstopar) [inline src/modules/partitioning/spectral.py]: You don't need the random clustering. That was only used for evaluation.

Comment (tinesubic) [reply src/modules/partitioning/spectral.py]: Fixed

File: src/modules/partitioning/k_means_sphere.py (hunk 1 above)
Comment (tinesubic) [reply src/modules/partitioning/k_means_sphere.py:104]: Thanks, I missed a `break` statement. Did you mean anything else? I've translated code from [this](https://github.com/lstopar/spectral-partitioning/blob/master/kMeansSphere.m) file

**Review by lstopar — APPROVED** [review]

**Pull request merged**